src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,997 @@
+  // that bitwise-not is a strictly decreasing function, if new_hi is the
+  // largest value not larger than hi that satisfies {bits._zeros, bits._ones},
+  // then ~new_hi is the smallest value not smaller than ~hi that satisfies
+  // {bits._ones, bits._zeros}.
+  //
+  // Proof:
+  // Calling h the smallest value not smaller than ~hi that satisfies
+  // {bits._ones, bits._zeros}.
+  //
+  // 1. Since h satisfies {bits._ones, bits._zeros}, ~h satisfies
+  //   {bits._zeros, bits._ones}. Assume the contradiction ~h does not satisfy
+  //   {bits._zeros, bits._ones}, There can be 2 cases:
+  //   1.1. There is a bit in ~h that is 0 where the corresponding bit in ones
+  //     is 1. This implies the corresponding bit in h is 1. But this is
+  //     contradictory since h satisfies {bits._ones, bits._zeros}.
+  //   1.2. There is a bit in ~h that is 1 where the corresponding bit in zeros
+  //     is 1. Similarly, this leads to contradiction because h needs to
+  //     satisfy {bits._ones, bits._zeros}.
+  //
+  // 2. Assume there is a value k that is larger than ~h such that k is not
+  // larger than hi and k satisfies {bits._zeros, bits._ones}. As a result, ~k
+  // would satisfy {bits._ones, bits._zeros}. And since bitwise-not is a
+  // strictly decreasing function, given ~h < k <= hi, we have h > ~k >= ~hi.
+  // This contradicts the assumption that h is the smallest value not smaller
+  // than ~hi and satisfies {bits._ones, bits._zeros}.
+  //
+  // As a result, ~h is the largest value not larger than hi that satisfies
+  // bits (QED).
+  U new_hi = ~adjust_lo(~bounds._hi, {bits._ones, bits._zeros});
+  if (new_hi > bounds._hi) {
+    return AdjustResult<RangeInt<U>>::make_empty();
+  }
+
+  bool progress = (new_lo != bounds._lo) || (new_hi != bounds._hi);
+  bool present = new_lo <= new_hi;
+  return {progress, present, {new_lo, new_hi}};
+}
+
+// Try to tighten the known bit constraints from the bound information by
+// extracting the common prefix of lo and hi and combining with the current
+// bit constraints
+// E.g: lo = 010011
+//      hi = 010100,
+// then all values in [lo, hi] would be
+//           010***
+template <class U>
+static AdjustResult<KnownBits<U>>
+adjust_bits_from_bounds(const KnownBits<U>& bits, const RangeInt<U>& bounds) {
+  // Find the mask to filter the common prefix, all values between bounds._lo
+  // and bounds._hi should share this common prefix in terms of bits
+  U mismatch = bounds._lo ^ bounds._hi;
+  // Find the first mismatch, all bits before it is the same in bounds._lo and
+  // bounds._hi
+  U match_mask = mismatch == U(0) ? std::numeric_limits<U>::max()
+                                  : ~(std::numeric_limits<U>::max() >> count_leading_zeros(mismatch));
+  // match_mask & bounds._lo is the common prefix, extract zeros and ones from
+  // it
+  U new_zeros = bits._zeros | (match_mask & ~bounds._lo);
+  U new_ones = bits._ones | (match_mask & bounds._lo);
```

Comment:
```suggestion
  U common_prefix = match_mask & bounds._lo;
  assert(common_prefix == match_mask & bounds._lo, "common prefix of both");
  U neg_common_prefix = match_mask & ~bounds._lo;
  assert(neg_common_prefix == match_mask & ~bounds._lo, "common negated prefix of both");
  U new_zeros = bits._zeros | neg_common_prefix;
  U new_ones = bits._ones | common_prefix;
```
Just an idea. Up to you.

Reply:
Hmm, I think your solution is understandable already, not sure mine is better. I'll still leave it for you here, feel free to mark it resolved.

Reply:
Nice suggestions, I have done so with a little modification.